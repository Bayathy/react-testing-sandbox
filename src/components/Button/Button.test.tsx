import { cleanup, fireEvent, render, screen } from "@testing-library/react"

import * as AlertContext from "~/context/AlertContext"

import { Button } from "./"

let alertDispatch: jest.SpyInstance<unknown>
let alertShowDispatchSpy: jest.SpyInstance<unknown>
let alertHideDispatchSpy: jest.SpyInstance<unknown>

describe("Button", () => {
  beforeEach(() => {
    alertDispatch = jest.spyOn(AlertContext, "useAlertDispatch")
    alertShowDispatchSpy = jest.fn()
    alertHideDispatchSpy = jest.fn()
    alertDispatch.mockImplementation(() => ({
      showDispatcher: alertShowDispatchSpy,
      hideDispatcher: alertHideDispatchSpy
    }))
  })
  afterEach(() => {
    alertDispatch.mockClear()
    alertShowDispatchSpy.mockClear()
    alertHideDispatchSpy.mockClear()
    cleanup()
  })
  test("render", () => {
    const { asFragment } = render(<Button />)
    expect(asFragment()).toMatchSnapshot()
  })
  test("click: show message dispatch", () => {
    const { container } = render(<Button />)
    const input = container.querySelector("input") as HTMLInputElement
    const button = container.querySelector("button") as HTMLButtonElement
    fireEvent.change(input, { target: { value: "test message" } })
    fireEvent.click(button)
    expect(alertShowDispatchSpy).toBeCalledWith("test message")
  })
  test("click: hide dispatch", () => {
    render(<Button />)
    const button = screen.getByText("Alert Close")
    fireEvent.click(button)
    expect(alertHideDispatchSpy).toBeCalled()
  })
})
