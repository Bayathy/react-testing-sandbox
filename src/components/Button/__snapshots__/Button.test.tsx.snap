// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button render 1`] = `
<DocumentFragment>
  <div>
    <input
      type="text"
      value=""
    />
    <br />
    <button>
      Alert
    </button>
    <br />
    <button>
      Alert Close
    </button>
  </div>
</DocumentFragment>
`;
